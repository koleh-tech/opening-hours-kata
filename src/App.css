#root {
    max-width: 1280px;
    margin: 0 auto;
    padding: 2rem;
    text-align: center;
    font-family: helvetica;
    font-size: 1.3rem;
}

.logo {
    height: 6em;
    padding: 1.5em;
    will-change: filter;
    transition: filter 300ms;
}
.logo:hover {
    filter: drop-shadow(0 0 2em #646cffaa);
}

input {
    border: none;
    font-size: inherit;
    font-family: inherit;
    border-radius: 5px;
    background-color: #6a0250;
    color: white;
}

label {
    font: inherit;
}

.configuration-option {
    display: flex;
    align-items: center;
    margin-bottom: 0.5rem;
    position: relative;
    left: 23%;
    gap: 0.5rem;
}

.configuration {
    border: 3px solid #5d5291;
    border-radius: 8px;
    background-color: #5d5291;
    color: white;
}

.card {
    padding: 2em;
}

.read-the-docs {
    color: #888;
}
