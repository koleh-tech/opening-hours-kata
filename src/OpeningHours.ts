const allDays = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]

class Day {
    constructor(
        public name: string,
        public isOpen: boolean,
    ) {}
}

export class ClosesBeforeOpeningError extends Error {
    public votingLink = "https://www.badgerbadgerbadger.com/"
    constructor() {
        super(
            "Closing the next day is not supported yet. Vote here if you'd like to see this feature:",
        )
    }
}

export class OpeningHours {
    private allDays: Day[]

    constructor(
        openDays: string[],
        private openingPeriod: Period,
    ) {
        this.allDays = allDays.map(
            (day) => new Day(day, openDays.includes(day)),
        )
    }

    openDayNames() {
        return this.allDays.filter((day) => day.isOpen).map((day) => day.name)
    }

    isOpenOn(input: Datetime) {
        return (
            this.openDayNames().includes(input.shortDayName()) &&
            this.openingPeriod.includes(input.asDate())
        )
    }

    nextOpeningDate(input: Datetime) {
        const daysAfterInput = [
            ...this.daysAfter(input),
            ...this.allDays, // loop back to the beginning of the week
        ]
        return input
            .incrementBy(daysAfterInput.findIndex((day) => day.isOpen) + 1)
            .toISOString()
    }

    daysAfter(input: Datetime) {
        return this.allDays.slice(this.weekIndexFor(input))
    }

    weekIndexFor(input: Datetime) {
        return (
            this.allDays.map((day) => day.name).indexOf(input.shortDayName()) +
            1
        )
    }
}

export class Datetime {
    constructor(
        private date: string,
        public localeFormat = "en-AU",
    ) {}

    static fromDate(date: Date) {
        return new Datetime(date.toISOString())
    }

    asDate() {
        return new Date(this.date)
    }

    shortDayName() {
        return this.asDate().toLocaleDateString(this.localeFormat, {
            weekday: "short",
        })
    }

    longDayName() {
        return this.asDate().toLocaleDateString(this.localeFormat, {
            weekday: "long",
        })
    }

    incrementBy(incrementBy: number) {
        const inputDay = this.asDate()
        inputDay.setDate(inputDay.getDate() + incrementBy)
        return inputDay
    }

    format() {
        return `${this.date} (${this.shortDayName()})`
    }

    time() {
        return this.asDate().toLocaleTimeString("en-GB", {
            hour: "2-digit",
            minute: "2-digit",
        })
    }
}

export class Time {
    constructor(
        public hour: number,
        public minute: number,
    ) {}

    asSeenOn(input: Date) {
        const date = new Date(input)
        date.setHours(this.hour, this.minute)
        date.setMinutes(this.minute)
        return date
    }

    static fromString(input: string) {
        return new Time(
            parseInt(input.split(":")[0]),
            parseInt(input.split(":")[1]),
        )
    }
}

export class Period {
    static fromStrings(openTimeOld: string, closeTimeOld: string) {
        return new Period(
            Time.fromString(openTimeOld),
            Time.fromString(closeTimeOld),
        )
    }

    constructor(
        private openTime: Time,
        private closeTime: Time,
    ) {
        if (this.openTime.hour > this.closeTime.hour)
            throw new ClosesBeforeOpeningError()

        if (this.openTime.hour === this.closeTime.hour)
            if (this.openTime.minute > this.closeTime.minute)
                throw new ClosesBeforeOpeningError()
    }

    includes(date: Date) {
        return (
            date >= this.openTime.asSeenOn(date) &&
            date < this.closeTime.asSeenOn(date)
        )
    }
}
