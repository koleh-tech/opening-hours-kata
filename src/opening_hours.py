from datetime import datetime, time, timedelta
import calendar


class OpeningHours:
    def __init__(self, days, open_time, close_time):
        self.days = [day.lower() for day in days]
        self.open_time = time.fromisoformat(open_time)
        self.close_time = time.fromisoformat(close_time)

    def isOpenOn(self, dt: str):
        local_dt = datetime.fromisoformat(dt.replace("Z", "+00:00"))
        weekday = calendar.day_name[local_dt.weekday()].lower()
        return (
            weekday[:3] in self.days
            and self.open_time <= local_dt.time() <= self.close_time
        )

    def nextOpeningDate(self, dt: str):
        current_dt = datetime.fromisoformat(dt.replace("Z", "+00:00"))
        for i in range(8):
            candidate = current_dt + timedelta(days=i)
            weekday = calendar.day_name[candidate.weekday()].lower()[:3]
            if weekday in self.days:
                opening_datetime = candidate.replace(
                    hour=self.open_time.hour,
                    minute=self.open_time.minute,
                    second=0,
                    microsecond=0,
                )
                if opening_datetime > current_dt:
                    return opening_datetime.isoformat() + "Z"
        return None
