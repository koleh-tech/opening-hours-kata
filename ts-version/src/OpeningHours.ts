const allDays = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]

class OpeningDay {
    constructor(
        public name: string,
        public isOpen: boolean,
    ) {}
}

export class Datetime {
    constructor(public date: string) {}

    dayName() {
        return new Date(this.date).toLocaleDateString("en-AU", {
            weekday: "short",
        })
    }

    incrementBy(incrementBy: number) {
        const inputDay = new Date(this.date)
        inputDay.setDate(inputDay.getDate() + incrementBy)
        return inputDay
    }

    format(input) {
        return `${this.date} (${this.dayName()})`
    }
}

export class OpeningHours {
    public days: OpeningDay[]
    constructor(
        inputDays: string[],
        public openTime: string,
        public closeTime: string,
    ) {
        this.days = allDays.map(
            (day) => new OpeningDay(day, inputDays.includes(day)),
        )
    }

    withinOpeningHours(datetime: string) {
        const time = new Date(datetime).getHours()
        return (
            time >= parseInt(this.openTime) && time < parseInt(this.closeTime)
        )
    }

    get openDays() {
        return this.days.filter((day) => day.isOpen)
    }

    get openDayNames() {
        return this.days.filter((day) => day.isOpen).map((day) => day.name)
    }

    isOpenOn(date: string) {
        return (
            this.openDayNames.includes(new Datetime(date).dayName()) &&
            this.withinOpeningHours(date)
        )
    }

    nextOpeningDate(input: Datetime) {
        const daysAfter = this.days.slice(
            this.days.map((day) => day.name).indexOf(input.dayName()) + 1,
        )
        const nextOpenDayIndex =
            [...daysAfter, ...this.days].findIndex((day) => day.isOpen) + 1
        const newDay = input.incrementBy(nextOpenDayIndex)

        return newDay.toISOString()
    }

    timeFor(date: string) {
        return new Date(date).toLocaleTimeString("en-AU", {
            hour: "2-digit",
            minute: "2-digit",
        })
    }
}
