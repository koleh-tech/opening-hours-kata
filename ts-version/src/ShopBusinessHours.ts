const allDays = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]

class OpeningDay {
    constructor(
        public name: string,
        public isOpen: boolean,
    ) {}
}

export class Datetime {
    constructor(public date: string) {}

    toDate() {
        return new Date(this.date)
    }

    dayName() {
        return this.toDate().toLocaleDateString("en-AU", {
            weekday: "short",
        })
    }

    incrementBy(incrementBy: number) {
        const inputDay = this.toDate()
        inputDay.setDate(inputDay.getDate() + incrementBy)
        return inputDay
    }

    format() {
        return `${this.date} (${this.dayName()})`
    }
}

export class OpeningHours {
    constructor(
        public openTime: string,
        public closeTime: string,
    ) {}

    withinOpeningHours(datetime: Datetime) {
        const time = new Date(datetime.toDate()).getHours()
        return (
            time >= parseInt(this.openTime) && time < parseInt(this.closeTime)
        )
    }
}

export class ShopBusinessHours {
    public allDays: OpeningDay[]
    constructor(
        openDays: string[],
        public openingHours: OpeningHours,
    ) {
        this.allDays = allDays.map(
            (day) => new OpeningDay(day, openDays.includes(day)),
        )
    }

    get openDays() {
        return this.allDays.filter((day) => day.isOpen)
    }

    get openDayNames() {
        return this.openDays.map((day) => day.name)
    }

    isOpenOn(date: string) {
        const input = new Datetime(date)
        return (
            this.openDayNames.includes(input.dayName()) &&
            this.openingHours.withinOpeningHours(input)
        )
    }

    nextOpeningDate(input: Datetime) {
        const daysAfter = this.allDays.slice(
            this.allDays.map((day) => day.name).indexOf(input.dayName()) + 1,
        )
        const nextOpenDayIndex =
            [...daysAfter, ...this.allDays].findIndex((day) => day.isOpen) + 1
        const newDay = input.incrementBy(nextOpenDayIndex)

        return newDay.toISOString()
    }

    timeFor(date: string) {
        return new Date(date).toLocaleTimeString("en-AU", {
            hour: "2-digit",
            minute: "2-digit",
        })
    }
}
