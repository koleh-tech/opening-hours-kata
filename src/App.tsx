import { useState } from "react"
import "./App.css"
import {
    ClosesBeforeOpeningError,
    Datetime,
    Day,
    OpeningHours,
    Period,
    Time,
} from "./OpeningHours"
import openingHoursLogo from "./assets/opening-hours.png"
import closingHoursLogo from "./assets/closing-hours.png"

function App() {
    const [datetimeToCheck, setDatetimeToCheck] = useState(
        new Datetime(new Date(Date.now()).toISOString()),
    )
    const [openingHours, setOpeningHours] = useState(
        new OpeningHours(["Mon", "Thu"], Period.fromStrings("08:00", "16:30")),
    )
    const [errorMessage, setErrorMessage] = useState(
        new ClosesBeforeOpeningError(""),
    )

    const openingTimeOptions = [
        {
            currentConfig: openingHours.openingPeriod.formatOpenTime(),
            handleChange: (e: { target: { value: string } }) => {
                try {
                    return setOpeningPeriod(
                        Period.fromStrings(
                            e.target.value,
                            openingHours.openingPeriod.formatCloseTime(),
                        ),
                    )
                } catch (error: any) {
                    if (error instanceof ClosesBeforeOpeningError) {
                        setErrorMessage(error)
                    }
                }
            },
            label: "Open",
        },
        {
            currentConfig: openingHours.openingPeriod.formatCloseTime(),
            handleChange: (e: { target: { value: string } }) =>
                setOpeningPeriod(
                    Period.fromStrings(
                        openingHours.openingPeriod.formatOpenTime(),
                        e.target.value,
                    ),
                ),
            label: "Close",
        },
    ]
    return (
        <>
            <div>
                {renderLogo()}
                <p>The business, on: </p>
            </div>
            <input
                type="datetime-local"
                value={datetimeToCheck.format()}
                onChange={handleDatetimeInput}
            ></input>{" "}
            <small>({datetimeToCheck.longDayName()})</small>
            <p>
                is {openingHours.isOpenOn(datetimeToCheck) ? "open" : "closed"},
                and opens again on{" "}
                {Datetime.fromDate(
                    openingHours.nextOpeningDate(datetimeToCheck),
                ).longDayName()}
            </p>
            {errorMessage.message !== "" ? (
                <div>
                    <p>{errorMessage.message}</p>
                    <a href={errorMessage.votingLink}>Vote here</a>
                </div>
            ) : (
                <p></p>
            )}
            <div className="configuration">
                <h3>Configure hours:</h3>
                {openingTimeOptions.map((option) => (
                    <div className="configuration-option">
                        <input
                            type="time"
                            value={option.currentConfig}
                            onChange={option.handleChange}
                        ></input>
                        <label>{option.label}</label>
                    </div>
                ))}

                <h3>Configure days:</h3>
                {openingHours.allDays.map((day, index) => (
                    <div key={day.name} className="configuration-option">
                        <input
                            type="checkbox"
                            id={`day-${index}`}
                            checked={day.isOpen}
                            onChange={() => setOpenDay(day)}
                        />
                        <label htmlFor={`day-${index}`}>{day.name}</label>
                    </div>
                ))}
            </div>
        </>
    )

    function retrieveDateIndex(index: number, toSplit: string) {
        return parseInt(toSplit.split("-")[index])
    }

    function handleDatetimeInput(event: { target: { value: string } }) {
        const toConvert = event.target.value
        const newDate = new Date()
        newDate.setFullYear(retrieveDateIndex(0, toConvert))
        newDate.setMonth(retrieveDateIndex(1, toConvert) - 1)
        newDate.setDate(retrieveDateIndex(2, toConvert))

        const newTime = Time.fromString(toConvert.split("T")[1])
        return setDatetimeToCheck(Datetime.fromDate(newTime.asSeenOn(newDate)))
    }

    function setOpeningPeriod(newPeriod: Period) {
        setOpeningHours(new OpeningHours(openingHours.openDays, newPeriod))
    }

    function setOpenDay(inputDay: Day) {
        return setOpeningHours(
            new OpeningHours(
                openingHours.allDays
                    .map((day) => ({
                        ...day,
                        isOpen:
                            day.name !== inputDay.name
                                ? day.isOpen // leave as is
                                : !day.isOpen,
                    }))
                    .filter((d) => d.isOpen)
                    .map((d) => d.name),
                openingHours.openingPeriod,
            ),
        )
    }

    function renderLogo() {
        return openingHours.isOpenOn(datetimeToCheck) ? (
            <a
                href="https://www.flaticon.com/free-icons/opening-hours"
                title="opening hours icons"
                className="logo"
            >
                <img src={openingHoursLogo} alt="Opening-hours-business-icon" />
            </a>
        ) : (
            <a
                href="https://www.flaticon.com/free-icons/closed"
                title="closed icons"
                className="logo"
            >
                <img src={closingHoursLogo} alt="Opening-hours-business-icon" />
            </a>
        )
    }
}

export default App
