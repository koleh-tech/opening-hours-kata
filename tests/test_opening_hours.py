from src.opening_hours import OpeningHours
import pytest
from approvaltests.approvals import verify


@pytest.fixture
def shop_hours():
    return OpeningHours(
        days=["mon", "wed", "fri"], open_time="08:00", close_time="16:00"
    )


def test_is_open_on(shop_hours: OpeningHours):
    results = {
        "wednesday": shop_hours.isOpenOn("2016-05-11T12:22:11.824Z"),
        "wednesday_after_hours": shop_hours.isOpenOn("2016-05-11T16:01:00.000Z"),
        "thursday": shop_hours.isOpenOn("2016-05-12T12:22:11.824Z"),
    }
    verify(str(results))


def test_next_opening_date(shop_hours: OpeningHours):
    next_open = shop_hours.nextOpeningDate("2016-05-11T12:22:11.824Z")
    verify(next_open)
