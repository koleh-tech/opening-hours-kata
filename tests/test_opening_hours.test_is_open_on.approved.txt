{'wednesday': True, 'wednesday_after_hours': False, 'thursday': False}
